services/versioncache/versioncache.go

```diff
@@ -0,0 +1,320 @@
+		if err != nil {
+			return err
+		}
+		if len(servicesVersions) == 0 {
+			results.CheckAfter = s.cfg.ServiceCheckInterval
+
+			return nil
+		}
+		if servicesVersions[0].NextCheckAt.After(time.Now()) {
+			results.CheckAfter = time.Until(servicesVersions[0].NextCheckAt)
+			if results.CheckAfter < s.cfg.MinCheckInterval {
+				results.CheckAfter = s.cfg.MinCheckInterval
+			}
+
+			return nil
+		}
+
+		results.NeedsUpdate = true
+		results.ServiceID = servicesVersions[0].ServiceID
+
+		service, err := models.FindServiceByID(tx.Querier, servicesVersions[0].ServiceID)
+		if err != nil {
+			return err
+		}
+		if service.ServiceType != models.MySQLServiceType {
+			return nil
+		}
+
+		pmmAgents, err := models.FindPMMAgentsForService(tx.Querier, servicesVersions[0].ServiceID)
+		if err != nil {
+			return err
+		}
+		if len(pmmAgents) == 0 {
+			return errors.Errorf("pmmAgent not found for service")
+		}
+		results.AgentID = pmmAgents[0].AgentID
+
+		// shift the next check time for this service, so, in case of versions fetch error,
+		// it will not loop in trying, but will continue with other services.
+		nextCheckAt := time.Now().UTC().Add(s.cfg.ServiceCheckInterval)
+		if _, err := models.UpdateServiceSoftwareVersions(tx.Querier, servicesVersions[0].ServiceID,
+			models.UpdateServiceSoftwareVersionsParams{NextCheckAt: &nextCheckAt},
+		); err != nil {
+			return err
+		}
+
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return results, nil
+}
+
+func softwareName(s agents.Software) (models.SoftwareName, error) {
+	var softwareName models.SoftwareName
+	switch software := s.(type) {
+	case *agents.Mysqld:
+		softwareName = models.MysqldSoftwareName
+	case *agents.Xtrabackup:
+		softwareName = models.XtrabackupSoftwareName
+	case *agents.Xbcloud:
+		softwareName = models.XbcloudSoftwareName
+	case *agents.Qpress:
+		softwareName = models.QpressSoftwareName
+	default:
+		return "", errors.Errorf("invalid software type %T", software)
+	}
+
+	return softwareName, nil
+}
+
+func (s *Service) updateVersions() (time.Duration, error) {
+	r, err := s.prepareUpdateVersions()
+	if err != nil {
+		return s.cfg.MinCheckInterval, err
+	}
+
+	if !r.NeedsUpdate {
+		return r.CheckAfter, nil
+	}
+
+	softwares := []agents.Software{&agents.Mysqld{}, &agents.Xtrabackup{}, &agents.Xbcloud{}, &agents.Qpress{}}
+	versions, err := s.v.GetVersions(r.AgentID, softwares)
+	if err != nil {
+		return s.cfg.MinCheckInterval, err
+	}
+	if len(versions) != len(softwares) {
+		return s.cfg.MinCheckInterval, errors.Errorf("slices length mismatch: versions len %d != softwares len %d",
+			len(versions), len(softwares))
+	}
+
+	svs := make([]models.SoftwareVersion, 0, len(softwares))
+	for i, software := range softwares {
+		name, err := softwareName(software)
+		if err != nil {
+			return s.cfg.MinCheckInterval, err
+		}
+
+		if versions[i].Error != "" {
+			s.l.Warnf("failed to get version of %q software: %s", name, versions[i].Error)
```

Comment:
no, because there could be the case when the error happens in some of the above statements, so we will have a constant loop of trying to update software versions for the same service (e.g. if it's broken).

Reply:
Yeah, correct, let's do it in error check above.

Reply:
I mean that versions for the service won't be updated until next day, it's better to check it again earlier.

Reply:
updated